### PR TITLE
Add new semi-private API functions to set per-packet-id custom handlers.

### PR DESCRIPTION
**Pull request by zoff99**
This is to prepare for ToxAV becoming independent of toxcore internal calls.

This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1424)

### CODE REVIEW

File: toxcore/tox.c [hunk 1]
```diff
@@ -2193,7 +2225,26 @@ bool tox_friend_send_lossy_packet(Tox *tox, uint32_t friend_number, const uint8_
...
-    tox->friend_lossy_packet_callback = callback;
+    /** start at PACKET_ID_RANGE_LOSSY_CUSTOM_START so ToxAV Packets are excluded if the old setter function is used */
+    for (uint8_t i = PACKET_ID_RANGE_LOSSY_CUSTOM_START; i <= PACKET_ID_RANGE_LOSSY_END; ++i) {
+        tox->friend_lossy_packet_callback_per_pktid[i] = callback;
+    }
+}
+
+void tox_callback_friend_lossy_packet_per_pktid(Tox *tox, tox_friend_lossy_packet_cb *callback, uint8_t pktid)
+{
+    // no error code returned in this function?
```
Comment (iphydf) [inline toxcore/tox.c:2236]: If you want, you can make it return an error. It's internal (toxav) only. Maybe log it somewhere. Maybe add an assert, because if this function errors, we have a code bug.

File: toxcore/tox.c [hunk 2]
```diff
@@ -2224,7 +2275,29 @@ bool tox_friend_send_lossless_packet(Tox *tox, uint32_t friend_number, const uin
...
+}
+
+void tox_callback_friend_lossless_packet_per_pktid(Tox *tox, tox_friend_lossless_packet_cb *callback, uint8_t pktid)
+{
+    // no error code returned in this function?
+
+    if (callback != nullptr) {
+        if ((pktid >= PACKET_ID_RANGE_LOSSLESS_CUSTOM_START) && (pktid <= PACKET_ID_RANGE_LOSSLESS_CUSTOM_END)) {
+            tox->friend_lossless_packet_callback_per_pktid[pktid] = callback;
+        } else if (pktid == PACKET_ID_MSI) {
```
Comment (iphydf) [inline toxcore/tox.c:2290]: The branches have the same code, so you can put `||` and make it a single if.

File: toxcore/tox.c [hunk 3]
```diff
@@ -2236,6 +2309,20 @@ void tox_self_get_dht_id(const Tox *tox, uint8_t *dht_id)
...
 }
 
+void tox_set_av_object(Tox *tox, void *object)
+{
+    lock(tox);
+    tox->toxav_object = object;
+    unlock(tox);
+}
+
+void tox_get_av_object(const Tox *tox, void **object)
```
Comment (iphydf) [inline toxcore/tox.c:2319]: This function should return the object-ptr, not set a pointer to the object-ptr.

File: toxcore/tox.c [hunk 4]
```diff
@@ -309,20 +312,46 @@ static void tox_conference_peer_list_changed_handler(Messenger *m, uint32_t conf
...
+    if (length < 1) {
+        // no error handling?
+        return;
+    }
+
     struct Tox_Userdata *tox_data = (struct Tox_Userdata *)user_data;
 
-    if (tox_data->tox->friend_lossless_packet_callback != nullptr) {
-        tox_data->tox->friend_lossless_packet_callback(tox_data->tox, friend_number, data, length, tox_data->user_data);
+    if (tox_data->tox->friend_lossless_packet_callback_per_pktid[data[0]] != nullptr) {
```
Comment (iphydf) [inline toxcore/tox.c:350]: Use packet_id instead of data[0].

File: toxcore/tox.c [hunk 5]
```diff
@@ -309,20 +312,46 @@ static void tox_conference_peer_list_changed_handler(Messenger *m, uint32_t conf
...
+        unlock(tox_data->tox);
+        tox_data->tox->friend_lossless_packet_callback_per_pktid[data[0]](tox_data->tox, friend_number, data, length,
+                tox_data->user_data);
+        lock(tox_data->tox);
```
Comment (iphydf) [inline toxcore/tox.c:354]: Make a separate PR for all the unlock/lock and non-recursive mutex changes.

File: toxcore/tox.c (hunk 1 above)
Comment (zoff99) [reply toxcore/tox.c:2236]: no

File: toxcore/tox.c (hunk 5 above)
Comment (zoff99) [reply toxcore/tox.c:354]: that slipped in there by accident

File: toxcore/tox.c (hunk 3 above)
Comment (zoff99) [reply toxcore/tox.c:2319]: return and lock won't work together. that's why i did it this way

**Comment by zoff99** [comment]
ah bazel config is missing the new .h file

**Comment by zugz** [comment]
Comments:

* the two error conditions in tox_friend_send_lossy_packet could be combined into one.

* there are extra brackets in various places, e.g.
`((pktid >= PACKET_ID_RANGE_LOSSY_START) && (pktid <= PACKET_ID_RANGE_LOSSY_END))`,
and a weird newline in tox_callback_friend_lossless_packet_per_pktid,
and another one in the documentation on tox_callback_friend_lossless_packet_per_pktid.

* the API documentation on tox_friend_send_lossy_packet should be
updated with the new range of packets it accepts.

**Comment by zugz** [comment]
Some suggested rewordings (from irc):

tox.api.h:2672-2673
"The first byte of data was not one of the permitted values; for lossy 
packets the first byte must be in the range 192-254, and for lossless 
packets it must be either 69 or in the range 160-191"

tox.api.h:2719
"The first byte of data must be either 69 or in the range 160-191"

**Comment by zoff99** [comment]
@zugz 
added in: https://github.com/TokTok/c-toxcore/pull/1431/commits/bbec6ae7ecfc81fdadede0e006e11ffe3427f63d
contained in PR: #1431
